{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computational Drug Discovery Tutorial\n",
    "\n",
    "Welcome to this hands-on notebook accompanying the blog tutorial.  \n",
    "We'll cover:\n",
    "\n",
    "1. **Environment Setup**  \n",
    "2. **Protein Preparation**  \n",
    "3. **Ligand Preparation**  \n",
    "4. **Docking with AutoDock Vina**  \n",
    "5. **Optional Mini Virtual Screening**  \n",
    "6. **Basic ML Model**  \n",
    "7. **Next Steps & Wrap-Up**  \n",
    "\n",
    "---\n",
    "\n",
    "## 1. Environment & Initial Setup\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python version: 3.11.11 | packaged by conda-forge | (main, Dec  5 2024, 14:17:24) [GCC 13.3.0]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_37105/2248172262.py\", line 15, in <module>\n",
      "    import rdkit\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/__init__.py\", line 23, in <module>\n",
      "    from rdkit.Chem.Draw import IPythonConsole\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/__init__.py\", line 18, in <module>\n",
      "    from rdkit import DataStructs\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/DataStructs/__init__.py\", line 13, in <module>\n",
      "    from rdkit.DataStructs import cDataStructs\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_37105/2248172262.py\", line 15, in <module>\n",
      "    import rdkit\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/__init__.py\", line 23, in <module>\n",
      "    from rdkit.Chem.Draw import IPythonConsole\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/__init__.py\", line 20, in <module>\n",
      "    from rdkit.Chem import rdchem\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_37105/2248172262.py\", line 15, in <module>\n",
      "    import rdkit\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/__init__.py\", line 23, in <module>\n",
      "    from rdkit.Chem.Draw import IPythonConsole\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/__init__.py\", line 24, in <module>\n",
      "    from rdkit.Chem.rdmolops import *\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_37105/2248172262.py\", line 15, in <module>\n",
      "    import rdkit\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/__init__.py\", line 23, in <module>\n",
      "    from rdkit.Chem.Draw import IPythonConsole\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/Draw/__init__.py\", line 20, in <module>\n",
      "    from rdkit.Chem import rdDepictor\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_37105/2248172262.py\", line 15, in <module>\n",
      "    import rdkit\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/__init__.py\", line 23, in <module>\n",
      "    from rdkit.Chem.Draw import IPythonConsole\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/Draw/__init__.py\", line 21, in <module>\n",
      "    from rdkit.Chem.Draw import rdMolDraw2D\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Skipped loading some Pytorch utilities, missing a dependency. No module named 'torch'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This module requires PyTorch to be installed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_37105/2248172262.py\", line 16, in <module>\n",
      "    import deepchem\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/__init__.py\", line 8, in <module>\n",
      "    import deepchem.data\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/__init__.py\", line 19, in <module>\n",
      "    from deepchem.data.data_loader import DataLoader\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/data_loader.py\", line 16, in <module>\n",
      "    from deepchem.feat import UserDefinedFeaturizer, Featurizer\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/__init__.py\", line 24, in <module>\n",
      "    from deepchem.feat.molecule_featurizers import AtomicCoordinates\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/__init__.py\", line 10, in <module>\n",
      "    from deepchem.feat.molecule_featurizers.mol2vec_fingerprint import Mol2VecFingerprint\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/mol2vec_fingerprint.py\", line 5, in <module>\n",
      "    from rdkit.Chem import AllChem\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/AllChem.py\", line 27, in <module>\n",
      "    from rdkit.Chem.rdDistGeom import *\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_37105/2248172262.py\", line 16, in <module>\n",
      "    import deepchem\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/__init__.py\", line 8, in <module>\n",
      "    import deepchem.data\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/__init__.py\", line 19, in <module>\n",
      "    from deepchem.data.data_loader import DataLoader\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/data_loader.py\", line 16, in <module>\n",
      "    from deepchem.feat import UserDefinedFeaturizer, Featurizer\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/__init__.py\", line 24, in <module>\n",
      "    from deepchem.feat.molecule_featurizers import AtomicCoordinates\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/__init__.py\", line 10, in <module>\n",
      "    from deepchem.feat.molecule_featurizers.mol2vec_fingerprint import Mol2VecFingerprint\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/mol2vec_fingerprint.py\", line 5, in <module>\n",
      "    from rdkit.Chem import AllChem\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/AllChem.py\", line 29, in <module>\n",
      "    from rdkit.Chem.rdMolAlign import *\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_37105/2248172262.py\", line 16, in <module>\n",
      "    import deepchem\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/__init__.py\", line 8, in <module>\n",
      "    import deepchem.data\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/__init__.py\", line 19, in <module>\n",
      "    from deepchem.data.data_loader import DataLoader\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/data_loader.py\", line 16, in <module>\n",
      "    from deepchem.feat import UserDefinedFeaturizer, Featurizer\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/__init__.py\", line 24, in <module>\n",
      "    from deepchem.feat.molecule_featurizers import AtomicCoordinates\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/__init__.py\", line 10, in <module>\n",
      "    from deepchem.feat.molecule_featurizers.mol2vec_fingerprint import Mol2VecFingerprint\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/mol2vec_fingerprint.py\", line 5, in <module>\n",
      "    from rdkit.Chem import AllChem\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/AllChem.py\", line 31, in <module>\n",
      "    from rdkit.Chem.rdMolTransforms import *\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_37105/2248172262.py\", line 16, in <module>\n",
      "    import deepchem\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/__init__.py\", line 8, in <module>\n",
      "    import deepchem.data\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/__init__.py\", line 19, in <module>\n",
      "    from deepchem.data.data_loader import DataLoader\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/data_loader.py\", line 16, in <module>\n",
      "    from deepchem.feat import UserDefinedFeaturizer, Featurizer\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/__init__.py\", line 24, in <module>\n",
      "    from deepchem.feat.molecule_featurizers import AtomicCoordinates\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/__init__.py\", line 10, in <module>\n",
      "    from deepchem.feat.molecule_featurizers.mol2vec_fingerprint import Mol2VecFingerprint\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/mol2vec_fingerprint.py\", line 5, in <module>\n",
      "    from rdkit.Chem import AllChem\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/AllChem.py\", line 33, in <module>\n",
      "    from rdkit.Chem.rdReducedGraphs import *\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_37105/2248172262.py\", line 16, in <module>\n",
      "    import deepchem\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/__init__.py\", line 8, in <module>\n",
      "    import deepchem.data\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/__init__.py\", line 19, in <module>\n",
      "    from deepchem.data.data_loader import DataLoader\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/data_loader.py\", line 16, in <module>\n",
      "    from deepchem.feat import UserDefinedFeaturizer, Featurizer\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/__init__.py\", line 24, in <module>\n",
      "    from deepchem.feat.molecule_featurizers import AtomicCoordinates\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/__init__.py\", line 10, in <module>\n",
      "    from deepchem.feat.molecule_featurizers.mol2vec_fingerprint import Mol2VecFingerprint\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/mol2vec_fingerprint.py\", line 5, in <module>\n",
      "    from rdkit.Chem import AllChem\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/AllChem.py\", line 34, in <module>\n",
      "    from rdkit.Chem.rdShapeHelpers import *\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_37105/2248172262.py\", line 16, in <module>\n",
      "    import deepchem\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/__init__.py\", line 8, in <module>\n",
      "    import deepchem.data\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/__init__.py\", line 19, in <module>\n",
      "    from deepchem.data.data_loader import DataLoader\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/data/data_loader.py\", line 16, in <module>\n",
      "    from deepchem.feat import UserDefinedFeaturizer, Featurizer\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/__init__.py\", line 24, in <module>\n",
      "    from deepchem.feat.molecule_featurizers import AtomicCoordinates\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/__init__.py\", line 22, in <module>\n",
      "    from deepchem.feat.molecule_featurizers.dmpnn_featurizer import DMPNNFeaturizer\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/dmpnn_featurizer.py\", line 24, in <module>\n",
      "    class GraphConvConstants(object):\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/dmpnn_featurizer.py\", line 57, in GraphConvConstants\n",
      "    RDKitDescriptors(use_bcut2d=False),\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/deepchem/feat/molecule_featurizers/rdkit_descriptors.py\", line 87, in __init__\n",
      "    from rdkit.Chem import Descriptors\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/Descriptors.py\", line 234, in <module>\n",
      "    _setupDescriptors(locals())\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/Descriptors.py\", line 31, in _setupDescriptors\n",
      "    from rdkit.Chem import GraphDescriptors, MolSurf, Lipinski, Fragments, Crippen, Descriptors3D\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/Chem/GraphDescriptors.py\", line 23, in <module>\n",
      "    from rdkit.ML.InfoTheory import entropy\n",
      "  File \"/home/joe/miniconda3/envs/deepchem-env/lib/python3.11/site-packages/rdkit/ML/InfoTheory/__init__.py\", line 4, in <module>\n",
      "    from rdkit.ML.InfoTheory.rdInfoTheory import *\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;31mAttributeError\u001b[0m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Skipped loading some Tensorflow models, missing a dependency. No module named 'tensorflow'\n",
      "Skipped loading some PyTorch models, missing a dependency. No module named 'torch'\n",
      "No module named 'torch'\n",
      "Skipped loading modules with pytorch-geometric dependency, missing a dependency. No module named 'torch'\n",
      "Skipped loading modules with pytorch-lightning dependency, missing a dependency. No module named 'torch'\n",
      "Skipped loading some Jax models, missing a dependency. No module named 'jax'\n",
      "Skipped loading some PyTorch models, missing a dependency. No module named 'tensorflow'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RDKit version: 2022.09.5\n",
      "DeepChem version: 2.8.0\n",
      "Data directory: ./data\n",
      "Results directory: ./results\n"
     ]
    }
   ],
   "source": [
    "# If running on Colab, install necessary packages here:\n",
    "# (Uncomment the lines below if you need them)\n",
    "\n",
    "# !pip install rdkit-pypi openbabel deepchem pyimolecule\n",
    "# !apt-get update\n",
    "# !apt-get install -y openbabel\n",
    "\n",
    "import sys\n",
    "import os\n",
    "\n",
    "# Check Python and library versions\n",
    "print(\"Python version:\", sys.version)\n",
    "\n",
    "try:\n",
    "    import rdkit\n",
    "    import deepchem\n",
    "    from rdkit import Chem\n",
    "    print(\"RDKit version:\", rdkit.__version__)\n",
    "    print(\"DeepChem version:\", deepchem.__version__)\n",
    "except ImportError as e:\n",
    "    print(\"Some packages are missing. Please install them before proceeding.\")\n",
    "    raise e\n",
    "\n",
    "# Define a data directory (adjust to your structure)\n",
    "DATA_DIR = \"./data\"\n",
    "RESULTS_DIR = \"./results\"\n",
    "\n",
    "os.makedirs(DATA_DIR, exist_ok=True)\n",
    "os.makedirs(RESULTS_DIR, exist_ok=True)\n",
    "\n",
    "print(\"Data directory:\", DATA_DIR)\n",
    "print(\"Results directory:\", RESULTS_DIR)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deepchem-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
